src/splitter.rs

```diff
@@ -15,8 +17,109 @@ impl Splitter {
+    pub fn phrases(&self, text: &str) -> Vec<Vec<u16>> {
+        // Split into sentences using delimiters: . ? \n\n ;
+        let sentences = self.split_into_sentences(text);
+        
+        // Collect all phrases (word sequences) from all sentences
+        let mut all_phrases = BTreeSet::new();
+        
+        for sentence in &sentences {
+            let cleaned_words = self.clean_sentence(sentence);
+            if cleaned_words.len() >= 2 {
+                // Generate all substrings of length >= 2
+                for phrase in self.generate_substrings(&cleaned_words) {
+                    all_phrases.insert(phrase);
+                }
+            }
+        }
+        
+        // Build vocabulary from the same cleaned text that we use for phrases
+        let vocabulary = self.build_cleaned_vocabulary(text);
+        
+        // Convert phrases to indices
+        all_phrases.into_iter()
+            .map(|phrase| self.phrase_to_indices(&phrase, &vocabulary))
+            .collect()
+    }
+    
+    fn split_into_sentences(&self, text: &str) -> Vec<String> {
+        // First split by \n\n, then split each part by other delimiters
+        let mut sentences = Vec::new();
+        
+        for paragraph in text.split("\n\n") {
+            for sentence in paragraph.split(|c| matches!(c, '.' | '?' | ';')) {
+                let trimmed = sentence.trim();
+                if !trimmed.is_empty() {
+                    sentences.push(trimmed.to_string());
+                }
+            }
+        }
+        
+        sentences
+    }
+    
+    fn clean_sentence(&self, sentence: &str) -> Vec<String> {
+        sentence
+            .chars()
+            .map(|c| {
+                // Keep 's as part of words, remove other punctuation
+                if c.is_alphabetic() || c.is_whitespace() || c == '\'' {
+                    c
+                } else {
+                    ' '
+                }
+            })
+            .collect::<String>()
+            .split_whitespace()
+            .map(|word| word.to_lowercase())
+            .filter(|word| !word.is_empty())
+            .collect()
+    }
+    
+    fn generate_substrings(&self, words: &[String]) -> Vec<Vec<String>> {
+        let mut substrings = Vec::new();
+        
+        // Generate all substrings of length >= 2
+        for start in 0..words.len() {
+            for end in (start + 2)..=words.len() {
+                substrings.push(words[start..end].to_vec());
+            }
+        }
+        
+        substrings
+    }
+    
+    fn phrase_to_indices(&self, phrase: &[String], vocabulary: &[String]) -> Vec<u16> {
+        phrase.iter()
+            .map(|word| {
+                vocabulary.iter()
+                    .position(|v| v == word)
+                    .expect("Word should be in vocabulary") as u16
+            })
+            .collect()
+    }
+    
+    fn build_cleaned_vocabulary(&self, text: &str) -> Vec<String> {
```

Comment:
Delete this - if there is something in vocabulary that isn't right for this case, feel free to edit vocabulary and its tests despite the issue saying only to edit this one method